src/v/ssx/single_fiber_executor.h

```diff
@@ -0,0 +1,324 @@
+      void,
+      stored_t>;
+    using wrapped_ret_value_t = std::expected<ret_value_t, errc>;
+    using promise_t = ss::promise<wrapped_ret_value_t>;
+    using ss::weakly_referencable<single_fiber_executor<Func>>::weak_from_this;
+
+private:
+    struct handle_t {
+        // invariant: we never store a handle with a fulfilled promise or with a
+        // triggered abort source
+        promise_t promise;
+        const ss::lw_shared_ptr<ss::abort_source> as;
+
+        explicit handle_t(promise_t&& p)
+          : promise(std::move(p))
+          , as(ss::make_lw_shared<ss::abort_source>()) {}
+        handle_t(const handle_t&) = delete;
+        handle_t& operator=(const handle_t&) = delete;
+        handle_t(handle_t&&) = delete;
+        handle_t& operator=(handle_t&&) = delete;
+        ~handle_t() = default;
+    };
+
+    struct running_t {
+        std::optional<handle_t> handle;
+
+        explicit running_t(promise_t&& promise)
+          : handle(std::move(promise)) {}
+        void run(Func&& func, ss::weak_ptr<self_t>&& executor) {
+            try {
+                ssx::background
+                  = func(*handle->as)
+                      .then_wrapped([executor = std::move(executor),
+                                     as = handle->as](future_t&& f) mutable {
+                          if (executor) {
+                              executor->complete_running(std::move(f));
+                          } else {
+                              // caller received errc::interrupted, no-one is
+                              // interested in the result
+                              f.ignore_ready_future();
+                          }
+                          return ss::now();
+                      });
+            } catch (...) {
+                vassert(executor, "executor is nullptr");
+                executor->complete_running(
+                  ss::make_exception_future<ret_value_t>(
+                    std::current_exception()));
+            }
+        }
+        running_t(const running_t&) = default;
+        running_t& operator=(const running_t&) = default;
+        running_t(running_t&&) = default;
+        running_t& operator=(running_t&&) = default;
+        ~running_t() { vassert(!handle, "promise not fulfilled"); };
+    };
+
+    struct requested_t {
+        Func func;
+        promise_t promise;
+
+        explicit requested_t(Func&& func)
+          : func(std::forward<Func>(func)) {}
+    };
+
+    // _running is set iff the background fiber is running
+    std::optional<running_t> _running;
+    std::optional<requested_t> _pending;
+
+    enum class status { operating, draining, drained };
+    status _status{status::operating};
+
+    void launch(promise_t&& promise, Func&& func) {
+        vassert(!_running, "launch() called while already running");
+        _running.emplace(std::move(promise));
+        _running->run(std::forward<Func>(func), weak_from_this());
+    }
+
+    void abort_running() {
```

Comment:
How about the case where abort running races with the completion of the running fiber? E.g., fiber A is running and about to finish and B is submitted. We call abort_running so the submitted of A gets `errc:interrupted` here, but A might then finish "successfully" without checking its abort source. So there is a mismatch between what the caller of A saw and what A did. 

Perhpas that's fine but it should be documented, i.e., the effects of A may have taken effect even if interrupted. Or maybe it's simply possible to avoid this state and the reasoning around it by requesting abort but not immediately competing the promise as interrupted but rather waiting to see the final result and/or if the running future observed the exception or not.

Reply:
Yes that's intended: we prefer to let the caller of the aborted function know it was interrupted ASAP. The whole idea of this runner is that the last function submitted totally devalues any previously submitted functions. I'll document this.